File: purchases/src/main/java/com/revenuecat/purchases/Dispatcher.java
package com.revenuecat.purchases;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.ExecutorService;


class Dispatcher {
    abstract static class AsyncCall implements Runnable {
        abstract public HTTPClient.Result call() throws HTTPClient.HTTPErrorException;
        void onError(int code, String message) {}
        void onCompletion(HTTPClient.Result result) {}

        @Override
        public void run() {
            Handler mainHandler = new Handler(Looper.getMainLooper());
            try {
                final HTTPClient.Result result = call();
                mainHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        onCompletion(result);
                    }
                });
            } catch (final HTTPClient.HTTPErrorException e) {
                mainHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        onError(0, e.getMessage());
                    }
                });
            }
        }
    }

    private ExecutorService executorService;

    Dispatcher(ExecutorService executorService) {
        this.executorService = executorService;
    }

    public void enqueue(AsyncCall call) {
        this.executorService.submit(call);
    }
    public void close() {
        this.executorService.shutdownNow();
    }
    public boolean isClosed() {
        return this.executorService.isShutdown();
    }
}

File: purchases/src/main/kotlin/com/revenuecat/purchases/Dispatcher.kt
//  Purchases
//
//  Copyright © 2019 RevenueCat, Inc. All rights reserved.
//

package com.revenuecat.purchases

import org.json.JSONException
import java.io.IOException
import java.util.concurrent.ExecutorService

internal open class Dispatcher(
    private val executorService: ExecutorService
) {

    internal abstract class AsyncCall : Runnable {
        @Throws(JSONException::class, IOException::class)
        abstract fun call(): HTTPClient.Result

        open fun onError(error: PurchasesError) {}
        open fun onCompletion(result: HTTPClient.Result) {}

        override fun run() {
            try {
                onCompletion(call())
            } catch (e: JSONException) {
                onError(e.toPurchasesError())
            } catch (e: IOException) {
                onError(e.toPurchasesError())
            } catch (e: SecurityException) {
                // This can happen if a user disables the INTERNET permission.
                onError(e.toPurchasesError())
            }
        }
    }

    open fun enqueue(call: AsyncCall) {
        this.executorService.execute(call)
    }

    open fun close() {
        this.executorService.shutdownNow()
    }

    open fun isClosed() = this.executorService.isShutdown
}
